In [2]:
import pandas as pd
import numpy as np


In [3]:
data_path = "football.csv"
df = pd.read_csv(data_path)
df.head()


Team  Goals  Shots on target  Shots off target Shooting Accuracy  \
0         Croatia      4               13                12             51.9%   
1  Czech Republic      4               13                18             41.9%   
2         Denmark      4               10                10             50.0%   
3         England      5               11                18             50.0%   
4          France      3               22                24             37.9%   

  % Goals-to-shots  Total shots (inc. Blocked)  Hit Woodwork  Penalty goals  \
0            16.0%                          32             0              0   
1            12.9%                          39             0              0   
2            20.0%                          27             1              0   
3            17.2%                          40             0              0   
4             6.5%                          65             1              0   

   Penalties not scored  ...  Saves made  Saves-to-shots ratio  Fouls Won  \
0                     0  ...          13                 81.3%         41   
1                     0  ...           9                 60.1%         53   
2                     0  ...          10                 66.7%         25   
3                     0  ...          22                 88.1%         43   
4                     0  ...           6                 54.6%         36   

  Fouls Conceded  Offsides  Yellow Cards  Red Cards  Subs on  Subs off  \
0             62         2             9          0        9         9   
1             73         8             7          0       11        11   
2             38         8             4          0        7         7   
3             45         6             5          0       11        11   
4             51         5             6          0       11        11   

   Players Used  
0            16  
1            19  
2            15  
3            16  
4            19  

[5 rows x 35 columns]

# Задача
Проанализировать набор данных `football.csv`, выполнив последовательность действий: импортировать необходимые библиотеки и загрузить данные в DataFrame pandas; получить сводную информацию о таблице с помощью методов `.info()`, `.describe()` и `.head()`; выбрать и вывести столбец `Goals`; посчитать общее число уникальных команд; определить количество строк и столбцов в наборе; создать датафрейм `discipline` из столбцов `Team`, `Yellow Cards`, `Red Cards`; отсортировать `discipline` по `Red Cards`, затем `Yellow Cards` по убыванию; вычислить среднее количество жёлтых карточек на команду; вывести команды, забившие более 6 голов; вывести данные команд, названия которых начинаются с буквы «G»; выбрать первые 7 столбцов исходного датафрейма; выбрать все столбцы, кроме последних трёх; определить точность ударов для Англии, Италии и России; найти TOP‑5 команд с максимальным и минимальным количеством нарушений (`Fouls Conceded`).

## Импорт библиотек и загрузка данных

### Subtask:
Импортировать необходимые библиотеки (pandas) и загрузить данные из файла football.csv в DataFrame.


**Обоснование**:
Необходимо импортировать библиотеку pandas и загрузить файл `football.csv` в датафрейм `df`.



In [4]:
print(f"Данные успешно загружены из {data_path}. Первые 5 строк:")
print(df.head())

Данные успешно загружены из football.csv. Первые 5 строк:
             Team  Goals  Shots on target  Shots off target Shooting Accuracy  \
0         Croatia      4               13                12             51.9%   
1  Czech Republic      4               13                18             41.9%   
2         Denmark      4               10                10             50.0%   
3         England      5               11                18             50.0%   
4          France      3               22                24             37.9%   

  % Goals-to-shots  Total shots (inc. Blocked)  Hit Woodwork  Penalty goals  \
0            16.0%                          32             0              0   
1            12.9%                          39             0              0   
2            20.0%                          27             1              0   
3            17.2%                          40             0              0   
4             6.5%                          65             1

## Сводная информация по датафрейму

### Subtask:
Получить сводную информацию о загруженном DataFrame, используя методы, такие как .info(), .describe(), .head().


**Обоснование**:
Нужно вывести сводную информацию о датафрейме с помощью методов `.info()`, `.describe()` и `.head()`. Удобно выполнить эти вызовы в одном кодовом блоке.



In [5]:
print("Сводка DataFrame:")
df.info()
print("\nОписательные статистики:")
print(df.describe())
print("\nПервые 5 строк (повторно):")
print(df.head())

Сводка DataFrame:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16 entries, 0 to 15
Data columns (total 35 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Team                        16 non-null     object 
 1   Goals                       16 non-null     int64  
 2   Shots on target             16 non-null     int64  
 3   Shots off target            16 non-null     int64  
 4   Shooting Accuracy           16 non-null     object 
 5   % Goals-to-shots            16 non-null     object 
 6   Total shots (inc. Blocked)  16 non-null     int64  
 7   Hit Woodwork                16 non-null     int64  
 8   Penalty goals               16 non-null     int64  
 9   Penalties not scored        16 non-null     int64  
 10  Headed goals                16 non-null     int64  
 11  Passes                      16 non-null     int64  
 12  Passes completed            16 non-null     int64  
 13  Passing Accuracy   

**Обоснование**:
Следующий шаг — выбрать и вывести столбец `Goals` из датафрейма. Для этого достаточно стандартного обращения к столбцу pandas.



In [6]:
print("Столбец Goals:")
print(df['Goals'])

Столбец Goals:
0      4
1      4
2      4
3      5
4      3
5     10
6      5
7      6
8      2
9      2
10     6
11     1
12     5
13    12
14     5
15     2
Name: Goals, dtype: int64


**Обоснование**:
Теперь нужно посчитать общее число уникальных команд. Для этого воспользуюсь столбцом `Team` и методом `.nunique()`.



In [7]:
unique_teams_count = df['Team'].nunique()
print(f"Общее число уникальных команд: {unique_teams_count}")

Общее число уникальных команд: 16


**Обоснование**:
Далее нужно определить количество строк и столбцов в наборе данных. Это удобно сделать через атрибут `.shape` датафрейма.



In [8]:
num_rows, num_cols = df.shape
print(f"Количество строк: {num_rows}")
print(f"Количество столбцов: {num_cols}")

Количество строк: 16
Количество столбцов: 35


**Обоснование**:
Теперь необходимо создать новый датафрейм `discipline`, включающий столбцы `Team`, `Yellow Cards` и `Red Cards`. Просто выберу эти колонки из исходного `df`.



In [9]:
discipline = df[['Team', 'Yellow Cards', 'Red Cards']]
print("Датафрейм discipline создан. Первые 5 строк:")
print(discipline.head())

Датафрейм discipline создан. Первые 5 строк:
             Team  Yellow Cards  Red Cards
0         Croatia             9          0
1  Czech Republic             7          0
2         Denmark             4          0
3         England             5          0
4          France             6          0


**Обоснование**:
Далее нужно отсортировать датафрейм `discipline` по столбцам `Red Cards`, затем `Yellow Cards` по убыванию. Для этого подойдёт метод `.sort_values()`.



In [10]:
discipline_sorted = discipline.sort_values(by=['Red Cards', 'Yellow Cards'], ascending=False)
print("discipline, отсортированный по Red Cards и Yellow Cards (по убыванию). Первые 5 строк:")
print(discipline_sorted.head())

discipline, отсортированный по Red Cards и Yellow Cards (по убыванию). Первые 5 строк:
                   Team  Yellow Cards  Red Cards
6                Greece             9          1
9                Poland             7          1
11  Republic of Ireland             6          1
7                 Italy            16          0
10             Portugal            12          0


**Обоснование**:
Нужно вычислить среднее количество жёлтых карточек на команду. Возьму столбец `Yellow Cards` и применю метод `.mean()`.



In [11]:
average_yellow_cards = df['Yellow Cards'].mean()
print(f"Среднее число жёлтых карточек на команду: {average_yellow_cards:.2f}")

Среднее число жёлтых карточек на команду: 7.44


**Обоснование**:
Далее требуется вывести команды, забившие более 6 голов. Для этого отфильтрую датафрейм `df` по условию для столбца `Goals`.



In [12]:
teams_more_than_6_goals = df[df['Goals'] > 6]
print("Команды, забившие более 6 голов:")
print(teams_more_than_6_goals[['Team', 'Goals']])

Команды, забившие более 6 голов:
       Team  Goals
5   Germany     10
13    Spain     12


**Обоснование**:
Затем нужно вывести данные команд, названия которых начинаются на букву «G». Использую строковый метод `str.startswith` для фильтрации по столбцу `Team`.



In [13]:
teams_starting_with_G = df[df['Team'].str.startswith('G')]
print("Команды, начинающиеся с буквы 'G':")
print(teams_starting_with_G[['Team', 'Goals']])

Команды, начинающиеся с буквы 'G':
      Team  Goals
5  Germany     10
6   Greece      5


**Обоснование**:
Теперь необходимо выбрать и показать первые 7 столбцов исходного датафрейма. Для этого воспользуюсь позиционным индексированием `iloc`.



In [14]:
first_7_columns = df.iloc[:, :7]
print("Первые 7 столбцов DataFrame:")
print(first_7_columns.head())

Первые 7 столбцов DataFrame:
             Team  Goals  Shots on target  Shots off target Shooting Accuracy  \
0         Croatia      4               13                12             51.9%   
1  Czech Republic      4               13                18             41.9%   
2         Denmark      4               10                10             50.0%   
3         England      5               11                18             50.0%   
4          France      3               22                24             37.9%   

  % Goals-to-shots  Total shots (inc. Blocked)  
0            16.0%                          32  
1            12.9%                          39  
2            20.0%                          27  
3            17.2%                          40  
4             6.5%                          65  


**Обоснование**:
Затем нужно вывести все столбцы, кроме последних трёх. Это также удобно сделать с помощью `iloc`, выбрав все строки и колонки до третьей с конца.



In [15]:
columns_except_last_3 = df.iloc[:, :-3]
print("Все столбцы, кроме последних трёх:")
print(columns_except_last_3.head())

Все столбцы, кроме последних трёх:
             Team  Goals  Shots on target  Shots off target Shooting Accuracy  \
0         Croatia      4               13                12             51.9%   
1  Czech Republic      4               13                18             41.9%   
2         Denmark      4               10                10             50.0%   
3         England      5               11                18             50.0%   
4          France      3               22                24             37.9%   

  % Goals-to-shots  Total shots (inc. Blocked)  Hit Woodwork  Penalty goals  \
0            16.0%                          32             0              0   
1            12.9%                          39             0              0   
2            20.0%                          27             1              0   
3            17.2%                          40             0              0   
4             6.5%                          65             1              0   

   

**Обоснование**:
Далее требуется определить точность ударов по воротам для Англии, Италии и России. Сначала преобразую столбец `Shooting Accuracy` в числовой формат, убрав символ `%` и приведя значения к `float`, затем отфильтрую нужные команды и выведу их показатели.



In [16]:
df['Shooting Accuracy'] = df['Shooting Accuracy'].str.replace('%', '').astype(float)

selected_countries = ['England', 'Italy', 'Russia']
shooting_accuracy_selected = df[df['Team'].isin(selected_countries)][['Team', 'Shooting Accuracy']]

print("Точность ударов Англии, Италии и России:")
print(shooting_accuracy_selected)

Точность ударов Англии, Италии и России:
       Team  Shooting Accuracy
3   England               50.0
7     Italy               43.0
12   Russia               22.5


**Обоснование**:
Теперь нужно определить TOP‑5 команд с наибольшим и наименьшим количеством нарушений. Сначала отсортирую датафрейм по `Fouls Conceded` по убыванию, чтобы получить лидеров по фолам.



In [17]:
top_5_most_fouls = df.sort_values(by='Fouls Conceded', ascending=False).head(5)
print("TOP-5 команд с наибольшим количеством фолов:")
print(top_5_most_fouls[['Team', 'Fouls Conceded']])

TOP-5 команд с наибольшим количеством фолов:
              Team  Fouls Conceded
10        Portugal              90
7            Italy              89
13           Spain              83
1   Czech Republic              73
0          Croatia              62


**Обоснование**:
После определения лидеров по нарушениям нужно вывести и команды с минимальным количеством фолов. Для этого отсортирую `df` по `Fouls Conceded` по возрастанию и возьму первые пять строк.



In [18]:
top_5_least_fouls = df.sort_values(by='Fouls Conceded', ascending=True).head(5)
print("TOP-5 команд с наименьшим количеством фолов:")
print(top_5_least_fouls[['Team', 'Fouls Conceded']])

TOP-5 команд с наименьшим количеством фолов:
           Team  Fouls Conceded
8   Netherlands              30
15      Ukraine              31
2       Denmark              38
12       Russia              43
3       England              45


## Итоги

### Основные результаты

* В наборе 16 уникальных команд, 16 строк и 35 столбцов; большинство колонок числовые (`int64`), пять имеют тип `object`, одна — `float64`.
* Среднее количество жёлтых карточек на команду составляет около 7.44.
* Более 6 голов забили только две команды: Испания (12) и Германия (10).
* Команды, название которых начинается с «G»: Germany (10 голов) и Greece (5 голов).
* Точность ударов по воротам для выбранных стран: Англия — 50.0%, Италия — 43.0%, Россия — 22.5%.
* TOP‑5 команд по количеству нарушений (`Fouls Conceded`): Португалия (90), Италия (89), Испания (83), Чехия (73), Хорватия (62).
* TOP‑5 наиболее дисциплинированных команд: Нидерланды (30), Украина (31), Дания (38), Россия (43), Англия (45).
* В датафрейме `discipline`, отсортированном по красным и жёлтым карточкам, по одной красной карточке получили Greece, Poland и Republic of Ireland; среди команд без удалений больше всего жёлтых карточек у Италии и Португалии.

### Наблюдения и возможные шаги

* **Дисциплина и результат:** Можно изучить связь между карточками и исходами матчей или количеством нарушений, чтобы понять влияние дисциплины на успех.
* **Сравнение метрик:** Сопоставление точности ударов с количеством голов поможет оценить эффективность и качество созданных моментов.
